### PR TITLE
[FLINK-28976][state/changelog] Don't add extra delay to the 1st materialization

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Spread changelog materialization and consequently deletion, more evenly in time.

To ease the creation of `PeriodicMaterializationManager` in test, this PR includes c7865f9663ffec5ca668a7781e2dc274ff7a35c9 from #19312.

## Verifying this change

- added `PeriodicMaterializationManagerTest.testInitialDelay`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ public void start() {
 
             LOG.info("Task {} starts periodic materialization", subtaskName);
 
-            scheduleNextMaterialization(periodicMaterializeDelay + initialDelay);
+            scheduleNextMaterialization(initialDelay);
```
Comment (rkhachatryan) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: This change affects the tests which use very high `periodicMaterializationDelay` to effictively disable materialization. There, `periodicMaterializationDelay` set to `Long.MAX` may or may not disable the materialization.

So we might want to (re)introduce a flag to disable materialization.

WDYT @curcur ?

Comment (fredia) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: > So we might want to (re)introduce a flag to disable materialization.
 
Currently,  if `PERIODIC_MATERIALIZATION_INTERVAL` is negative, the periodic materialization will be disabled. Do you mean to introduce a **new** flag?

Comment (curcur) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: I think we should have some way to disable the materialization (flag).

I am not a fan of using `periodicMaterializationDelay` very long to disable materialization. It is implicit and may lead error.

Comment (masteryhx) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: I think disabling materialization for test is not the main purpose of this pr, right ?
(Diabling has been supported in [FLINK-26673](https://issues.apache.org/jira/browse/FLINK-26673))
About making the checkpoint deletion evenly, It is not enough for users to using changlog out of box without configuring other configs like [jobmanager.io-pool.size](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/config/#jobmanager-io-pool-size) ?
We may still need other strategy to solve the problem of the checkpoint deletion delay ?

Comment (curcur) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: So now we have `periodicMaterializationDelay`, we have `initialDelay` and we have `periodicMaterializationDelay < 0` to indicate materialization disabled if I understand correctly?

Isn't this so confusing and misleading to the users?  What about `initialDelay` being very long or negative?

What is the problem to introduce a flag to indicate enable/disable? I think the original argument is that we do not want to introduce too many configs, but I think the current way is much more confusing and problematic.

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:121]: Sorry for the confusion, a negative value will prevent any materialization indeed.
I think both approaches have their both pros and cons (a separate flag is more explicit, but the user has to keep the two variables consistent).
But keeping negative value as it is doesn't change the API, so I'd chose this option.

File: flink-state-backends/flink-statebackend-common/src/test/java/org/apache/flink/state/common/PeriodicMaterializationManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.Test;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.apache.flink.shaded.guava30.com.google.common.collect.Iterators.getOnlyElement;
+import static org.apache.flink.util.concurrent.Executors.newDirectExecutorService;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+/** {@link PeriodicMaterializationManager} test. */
+public class PeriodicMaterializationManagerTest {
```
Comment (fredia) [inline flink-state-backends/flink-statebackend-common/src/test/java/org/apache/flink/state/common/PeriodicMaterializationManagerTest.java:34]: nit: extends TestLogger?

**Comment by fredia** [comment]
Thanks for creating this PR @rkhachatryan, LGTM. 
I noticed you mentioned this PR under [FLINK-26590](https://issues.apache.org/jira/browse/FLINK-26590?focusedCommentId=17579828&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17579828) :
> I found that the problem is caused by [FLINK-28976](https://issues.apache.org/jira/browse/FLINK-28976) in some setups.
The fix for which is much more simple, so I'd prioritize it over this ticket.

Why **removing `materialization_interval`** can solve FLINK-26590?  
I guess “removing `materialization_interval`” makes the size of 1st materialization smaller, thus the deletion of shared state can be faster?

**Review by fredia — APPROVED** [review]

**Comment by rkhachatryan** [comment]
Thanks for the review @fredia,

> Why removing materialization_interval can solve [FLINK-26590](https://issues.apache.org/jira/browse/FLINK-26590)?
> I guess “removing materialization_interval” makes the size of 1st materialization smaller, thus the deletion of shared state can be faster?

Yes, exactly.

**Pull request merged**